added public_gpg_key to user page

SydneyDlh opened the pull request:
added the public_gpg_key to the user page for users that have a public_gpg_key. If they have a key the page shows a preview of the first 50 characters and can be clicked which will download the key.

maorfr commented:
can we get this one reviewed by @jmelis please? :)

maorfr commented on src/pages/elements/User.js:
```diff
@@ -2,7 +2,32 @@ import React from 'react';
+      const element = document.createElement("a");
+      const file = new Blob([user.public_gpg_key], {type: 'text/plain'});
+      element.href = URL.createObjectURL(file);
+      element.download = `${user.redhat_username}_public_gpg_key.asc`;
```
since the keys in app-interface are binary base64 encoded and not ascii armored, `.asc` is probably a bit misleading. i think that `.gpg` would make more sense. what do you think?

maorfr commented on src/pages/elements/User.js:
```diff
@@ -19,11 +44,14 @@ function User({ user }) {
       />
-
       <h4>Roles</h4>
+      console.log(key);
```
debugging leftovers?

maorfr commented on src/pages/elements/User.js:
```diff
@@ -2,7 +2,32 @@ import React from 'react';
+    key = "-";
+  }
+  else if (user.public_gpg_key.length>= 50) {
+     key = <button onClick={downloadKey}> {user.public_gpg_key.substring(0,50) + '...'} </button>;
```
i like this a lot!
it makes so much sense that there will be a download button.
but i found it less intuitive then i thought.

maybe we can have something more similar to `Show more...` and `Show less...` and a download button?
or something like that...
i'll leave that for you to decide :)

SydneyDlh replied:
I wanted to do this, the way I would normally do this in react would be with a state change, but this is a function not a class so it does not have states (as far as I know). The files are structured differently then I've seen before which is what I wanted to ask @jmelis about. I can look more into it and see if I can figure out how to do state changes with how these files are set up.

maorfr commented:
@SydneyDlh, i really like that you were able to produce something working so soon. good job!

SydneyDlh commented:
I fixed the changes @maorfr  requested as well as adding a button to show more of the key and show less of the key. In addition I corrected all lint errors using eslint-config-airbnb as @jmelis requested. Let me know if you have any feedback!

maorfr commented:
this is starting to look really good, @SydneyDlh!
super nice that you found how to be able to do "show more"/"show less".

something is strange though. when i look at a user without a gpg key, the dash is one line below the `Public gpg Key` (btw, Public GPG key).
and when i click on "show more" - the entire layout changes.
plus, i expected that the buttons below the key...

i was thinking, can we replace those buttons with something that is a bit more similar to the entire layout?

thanks for all your hard work!

jmelis commented:
thanks for the review @maorfr, I'll wait for @SydneyDlh to address those comments to review it

SydneyDlh commented:
I changed the show more feature to use a ClipboardCopy a component from patternfly 4. This allows the user to expand a large amount of text (like the public GPG key) and copy it to clipboard.

This solved a few of the problems @maorfr mentioned, the layout no longer changes, and I think it matches the overall layout a little better.

It still needs to be styled but I was having some trouble with patternfly, will look more into it. Not sure if this is the correct solution since we still use patternfly 3 for everything else and because it isn't perfect (for example the expanded text repeats the first line shown in the preview).

jmelis commented on package.json:
```diff
@@ -3,6 +3,7 @@
   "version": "0.0.0",
   "private": true,
   "dependencies": {
+    "@patternfly/react-core": "^3.16.10",
```
:+1:  to start including patternfly 4 components

jmelis commented on src/pages/elements/User.js:
```diff
@@ -1,8 +1,72 @@
+      </div>
+    );
+  }
+  const [key, changeKey] = useState(keyState);
```
linter and compiler are complaining with:

```
src/pages/elements/User.js
  35:10  error  'key' is assigned a value but never used  no-unused-vars
```

perhaps we could just do this:

```diff
--- a/src/pages/elements/User.js
+++ b/src/pages/elements/User.js
@@ -32,7 +32,7 @@ function User({ user }) {
       </div>
     );
   }
-  const [key, changeKey] = useState(keyState);
+  const changeKey = useState(keyState)[1];
   const [showMoreButton, changeButton] = useState(showMoreKey);
 
   // taken from: https://stackoverflow.com/questions/44656610/download-a-string-as-txt-file-in-react
```

SydneyDlh commented:
@jmelis I fixed the compiler warning and styled the download button so it matched the rest of it better. Let me know what you think!
